### PR TITLE
Fix map update with incompatible ZIP extra data

### LINKED ISSUE #21323: Support packing maps to .oramap using third-party software
## System Information

- Operating System:  Windows 11 Pro (x64)
- OpenRA Version: DEV-VERSION (bleed)
- OpenRA Mod: ra
- Source: Official download package OR self-compiled OR third-party package
- For self-compiled or third-party packages: Mono version

## Exception log

```
RemoveValidRelationsFromCapturable: Remove ValidRelations property from Capturable.
   Updating map... FAILED

   The automated changes for this rule were not applied because of an error.
   After the issue reported below is resolved you should run the updater
   with SOURCE set to RemoveValidRelationsFromCapturable to retry these changes

   The exception reported was:
     ICSharpCode.SharpZipLib.Zip.ZipException: Failed to copy bytes expected 108 read 0
        at ICSharpCode.SharpZipLib.Zip.ZipFile.CopyEntryDataDirect(ZipUpdate update, Stream stream, Boolean updateCrc, Int64& destinationPosition, Int64& sourcePosition)
        at ICSharpCode.SharpZipLib.Zip.ZipFile.CopyEntryDirect(ZipFile workFile, ZipUpdate update, Int64& destinationPosition)
        at ICSharpCode.SharpZipLib.Zip.ZipFile.RunUpdates()
        at ICSharpCode.SharpZipLib.Zip.ZipFile.CommitUpdate()
        at OpenRA.FileSystem.ZipFileLoader.ReadWriteZipFile.Update(String filename, Byte[] contents) in D:\Project\OpenRA\OpenRA.Game\FileSystem\ZipFile.cs:line 128
        at OpenRA.Mods.Common.UpdateRules.UpdateExtensions.Save(List`1 files) in D:\Project\OpenRA\OpenRA.Mods.Common\UpdateRules\UpdateUtils.cs:line 349
        at OpenRA.Mods.Common.UtilityCommands.UpdateMapCommand.ApplyRules(ModData modData, IReadWritePackage mapPackage, IEnumerable`1 rules) in D:\Project\OpenRA\OpenRA.Mods.Common\UtilityCommands\UpdateMapCommand.cs:line 112
ExtractResourceStorageFromHarvester: Renames StoresResources to StoresPlayerResources and extracts StoresResources from Harvester.
   Updating map... FAILED

   The automated changes for this rule were not applied because of an error.
   After the issue reported below is resolved you should run the updater
   with SOURCE set to ExtractResourceStorageFromHarvester to retry these changes

   The exception reported was:
     System.ObjectDisposedException: Cannot access a disposed object.
     Object name: 'ZipFile'.
        at ICSharpCode.SharpZipLib.Zip.ZipFile.GetEntry(String name)
        at OpenRA.FileSystem.ZipFileLoader.ReadOnlyZipFile.GetStream(String filename) in D:\Project\OpenRA\OpenRA.Game\FileSystem\ZipFile.cs:line 40
        at OpenRA.Mods.Common.UpdateRules.UpdateUtils.UpdateMap(ModData modData, IReadWritePackage mapPackage, UpdateRule rule, List`1& files, HashSet`1 externalFilenames) in D:\Project\OpenRA\OpenRA.Mods.Common\UpdateRules\UpdateUtils.cs:line 88
        at OpenRA.Mods.Common.UtilityCommands.UpdateMapCommand.ApplyRules(ModData modData, IReadWritePackage mapPackage, IEnumerable`1 rules) in D:\Project\OpenRA\OpenRA.Mods.Common\UtilityCommands\UpdateMapCommand.cs:line 109
ReplacePaletteModifiers: Replace palette modifiers with post-processing shaders.
   Updating map... FAILED
```

## Replay
try to update the map https://resource.openra.net/maps/55706/

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/21323

### CODE REVIEW

File: OpenRA.Game/FileSystem/ZipFile.cs [hunk 1]
```diff
@@ -136,6 +137,12 @@ public void Delete(string filename)
 				pkg.CommitUpdate();
 				Commit();
 			}
+
+			public void ClearExtraData()
+			{
+				foreach (ZipEntry entry in pkg)
```
Comment (pchote) [inline OpenRA.Game/FileSystem/ZipFile.cs]: Probably best to do this inline in the ctor instead.

**Review by PunkPun — COMMENTED** [review]
I'm not sure this works. I tried using the default macOS zipper and it still can't read it.

I get `System.IO.FileNotFoundException: Required file map.yaml not present in this map`

[af_patching.zip](https://github.com/user-attachments/files/16414152/af_patching.zip)

version with sanitisation run. Same crash

[af_patching.zip](https://github.com/user-attachments/files/16414393/af_patching.zip)

**Comment by pchote** [comment]
That zip file contains a directory `af_patching` that contains `map.yaml` etc, which is invalid with or without this fix!

File: OpenRA.Game/FileSystem/ZipFile.cs [hunk 2]
```diff
@@ -114,6 +114,9 @@ public ReadWriteZipFile(string filename, bool create = false)
 				pkgStream.Position = 0;
 				pkg = new ZipFile(pkgStream);
 				Name = filename;
+
+				foreach (ZipEntry entry in pkg)
```
Comment (pchote) [inline OpenRA.Game/FileSystem/ZipFile.cs:119]: I'm sure that future maintainers would appreciate a comment here explaining why this is needed.

**Review by pchote — APPROVED** [review]

**Review by PunkPun — COMMENTED** [review]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2a3bc9ce883f6808e55120b828413ba2f07502d8...3f84b659819614548cfa3e6cb999f01e183bab20)

**Pull request merged**